codec-dns-doh/src/main/java/io/netty/handler/codec/dns/doh/DoHDecoder.java

```diff
@@ -0,0 +1,159 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.dns.doh;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.CorruptedFrameException;
+import io.netty.handler.codec.DecoderException;
+import io.netty.handler.codec.MessageToMessageDecoder;
+import io.netty.handler.codec.UnsupportedMessageTypeException;
+import io.netty.handler.codec.base64.Base64;
+import io.netty.handler.codec.base64.Base64Dialect;
+import io.netty.handler.codec.dns.DefaultDnsQuery;
+import io.netty.handler.codec.dns.DefaultDnsResponse;
+import io.netty.handler.codec.dns.DnsOpCode;
+import io.netty.handler.codec.dns.DnsQuery;
+import io.netty.handler.codec.dns.DnsRecord;
+import io.netty.handler.codec.dns.DnsRecordDecoder;
+import io.netty.handler.codec.dns.DnsResponse;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.handler.codec.dns.DnsSection;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.util.CharsetUtil;
+
+import java.net.SocketAddress;
+import java.net.URL;
+import java.util.List;
+
+/**
+ * Decode {@link DnsQuery} from {@link FullHttpRequest} or {@link FullHttpResponse}.
+ */
+public final class DoHDecoder extends MessageToMessageDecoder<HttpObject> {
+
+    private final DnsResponseDecoder<SocketAddress> dnsResponseDecoder;
+    private final DnsRecordDecoder recordDecoder = DnsRecordDecoder.DEFAULT;
+
+    public DoHDecoder() {
+        this.dnsResponseDecoder = new DnsResponseDecoder<SocketAddress>(recordDecoder) {
+            @Override
+            protected DnsResponse newResponse(SocketAddress sender, SocketAddress recipient, int id,
+                                              DnsOpCode opCode, DnsResponseCode responseCode) {
+                return new DefaultDnsResponse(id, opCode, responseCode);
+            }
+        };
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, HttpObject msg, List<Object> out) throws Exception {
+
+        if (msg instanceof FullHttpResponse) {
+            FullHttpResponse fullHttpResponse = (FullHttpResponse) msg;
+            int statusCode = fullHttpResponse.status().code();
+
+            if (!(statusCode >= 200 && statusCode <= 299)) {
+                throw new DecoderException("Request failed, received HTTP Response Code: " + statusCode);
+            }
+
+            out.add(dnsResponseDecoder.decode(ctx.channel().remoteAddress(), ctx.channel().localAddress(),
```

Comment:
It seems like you are ignoring the cache headers which the RFC documents cases when they are relevant to processing.  e.g. Age header.

Are these concerns out of scope of the initial PR?

https://tools.ietf.org/html/rfc8484#section-5.1

Reply:
Cache Headers are useful if we have some HTTP Caching layer. We're not using any caching here at all and people are free to use custom caching solution for native DnsQuery. So caching is not a concern for now.

Reply:
The decoder is currently incomplete and broken if caching is implemented by any server. You may want to add "no-cache" to the encoder to make it clear you have skipped any caching for now.

```
DoH clients MUST account for the Age response header field's value
   [RFC7234] when calculating the DNS TTL of a response.  For example,
   if an RRset is received with a DNS TTL of 600, but the Age header
   field indicates that the response has been cached for 250 seconds,
   the remaining lifetime of the RRset is 350 seconds.  This requirement
   applies to both DoH client HTTP caches and DoH client DNS caches.

   DoH clients can request an uncached copy of a HTTP response by using
   the "no-cache" request Cache-Control directive (see Section 5.2.1.4
   of [RFC7234]) and similar controls.  Note that some caches might not
   honor these directives, either due to configuration or interaction
   with traditional DNS caches that do not have such a mechanism.
```

Reply:
Makes sense now.